<div align='center'>
    <h1>Tugas Pendahuluan Praktikum Dasar Kecerdasan Artifisial</h1>
    <h3>Modul 10: Implementasi Simulated Annealing</h3>
</div>

In [1]:
print("Muhammad Alvin Faa'iz") # Tuliskan nama lengkap Anda
print("103012400229") # Tuliskan NIM Anda
print("IF-48-11") # Tuliskan kelas Anda
print("VVN") # Tuliskan kode asprak Anda

Muhammad Alvin Faa'iz
103012400229
IF-48-11
VVN


### Pengantar Simulated Annealing

In [2]:
import random
import math
import numpy as np # import library numpy

# Soal

Kamu diberi daftar lokasi pelanggan di beberapa kota. Tujuanmu adalah menemukan urutan pengiriman yang paling efisien agar kendaraan bisa menghemat bahan bakar.

Berikut adalah koordinat kota yang perlu dikunjungi:
- A: (2, 3)
- B: (5, 4)
- C: (1, 8)
- D: (7, 2)

Tugasmu adalah mencari urutan kota dengan jarak tempuh total paling minimum.

In [3]:
cities = {'A': (2, 3), 'B': (5, 4), 'C': (1, 8), 'D': (7, 2)}
route = ['A', 'B', 'C', 'D']

Buat fungsi untuk menghitung jarak Euclidean antara dua titik.

```math
d(x₁, y₁, x₂, y₂) = \sqrt{(x₂ - x₁)^2 + (y₂ - y₁)^2}
````

Pro tips: gunakan `math.sqrt()` untuk menghitung akar.

In [4]:
def euclidean_distance(p1, p2):
    return math.sqrt((p2[0] - p1[0])**2 + (p2[1] - p1[1])**2) # buat algoritma menghitung jarak Euclidean antara dua titik p1 dan p2

In [5]:
print(euclidean_distance((2,3), (5,4)))

3.1622776601683795


Contoh *output*

|input|output|
|---|---|
||3.1622776601683795|

Gunakan fungsi sebelumnya untuk menghitung total jarak dari urutan kota tertentu.

In [6]:
def total_distance(route, cities):
    distance = 0 # inisialisasi variabel distance dengan 0
    for i in range(len(route) - 1): # looping dari 0 sampai panjang route - 1
        distance += euclidean_distance(cities[route[i]], cities[route[i + 1]]) # lakukan increment pada variabel distance dengan memanggil function ecuclidean_distance
    return distance # mengembalikan nilai distance


In [7]:
print(total_distance(route, cities))

17.30441328389933


Contoh *output*

|input|output|
|---|---|
||17.30441328389933|

Gunakan algoritma Simulated Annealing untuk mencari rute optimal.

In [8]:
def simulated_annealing(cities, initial_temp=100, cooling_rate=0.95, iteration=100):
    current_solution = list(cities.keys())  # inisialisasi solusi awal dengan list dari kunci cities
    best_solution = current_solution[:]  # salin solusi awal sebagai solusi terbaik sementara
    current_distance = total_distance(current_solution, cities)  # hitung total jarak dari current_solution
    best_distance = current_distance  # set best_distance sama dengan current_distance
    temp = initial_temp  # set suhu awal dari parameter

    for _ in range(iteration):  # lakukan iterasi sebanyak nilai 'iteration'
        i, j = random.sample(range(len(current_solution)), 2)  # ambil dua index secara acak
        new_solution = current_solution[:]  # salin solusi saat ini
        new_solution[i], new_solution[j] = new_solution[j], new_solution[i]  # tukar dua kota dalam rute
        new_distance = total_distance(new_solution, cities)  # hitung total jarak dari new_solution

        # jika solusi baru lebih baik atau lolos probabilitas perubahan solusi
        if new_distance < current_distance or random.random() < math.exp((current_distance - new_distance) / temp):
            current_solution = new_solution[:]  # terima solusi baru sebagai current_solution
            current_distance = new_distance  # update current_distance
            if current_distance < best_distance:  # periksa apakah solusi saat ini lebih baik dari yang terbaik
                best_solution = current_solution[:]  # update best_solution
                best_distance = current_distance  # update best_distance

        temp *= cooling_rate  # kurangi suhu sesuai cooling_rate

    return best_solution, best_distance  # kembalikan best_solution dan best_distance


Lakukan uji algoritma yang telah dibuat dengan memanggil algoritma simualated annealing

In [12]:
best_solution, best_distance= simulated_annealing(cities)  # panggil fungsi simulated_annealing dengan parameter cities dan simpan hasilnya ke dua variabel

print("Rute terbaik: ", best_solution)  # tampilkan hasil rute terbaik
print("Jarak tempuh: ", best_distance)  # tampilkan jarak tempuh dari rute terbaik

Rute terbaik:  ['D', 'B', 'A', 'C']
Jarak tempuh:  11.089724298507354


Contoh *output*

|input|output|
|---|---|
||Rute terbaik:  ['C', 'A', 'B', 'D']<br>Jarak tempuh:  11.089724298507354|